File: 004_DeleteNodInLinkedList/test40.cpp
/**
 * @file test40.cpp
 * @author M. Z.
 * @brief Deleting a node in the linked list
 * @version 0.1
 * @date 2025-03-03
 *
 * @copyright Copyright (c) 2025
 *
 */
#include <iostream>

using namespace std;

struct Node {
  int data; // char data;
  Node *next;
};

Node *head; /*Global variable, can be accessed anywhere*/

void insertNodeInList(int data, int position); /*At the end of the list*/
void printList();
void deleteNodeInList(int position);

int main() {
  head = NULL; /*Empty list*/
  insertNodeInList(9, 1);
  insertNodeInList(8, 2);
  insertNodeInList(7, 3);
  insertNodeInList(6, 4);
  insertNodeInList(5, 5);
  printList();
  int position;
  cout << "Enter a position to be deleted: ";
  cin >> position;
  deleteNodeInList(position);
  printList();
}

void insertNodeInList(int data, int position) {
  Node *temp1 = new Node();
  temp1->data = data; // Similar to (*temp1).data = x;
  if (1 == position) {
    temp1->next = head;
    head = temp1;
  } else {
    Node *temp2 = head;
    for (int i = 0; i < position - 2; i++) {
      temp2 = temp2->next;
    }
    temp1->next = temp2->next;
    temp2->next = temp1;
  }
}

void printList() {
  Node *temp = head;
  cout << "List is: ";
  while (temp != NULL) {
    cout << " " << temp->data;
    temp = temp->next;
  }
  cout << endl;
}

void deleteNodeInList(int position) {
  Node *temp1 = head;
  if (position == 1) {
    head = temp1->next; /*head now points to second node*/
    delete temp1;
    return;
  }
  for (int i = 0; i < position - 2; i++) {
    temp1 = temp1->next;
  }
  /*temp1 points to (position-1)th node*/
  Node *temp2 = temp1->next; /*temp2 at position node*/
  temp1->next = temp2->next; /*(position + 1) node*/
  delete temp2;
}

File: 004_DeleteNodInLinkedList/test40.py
'''
 * @file test40.py
 * @author M. Z.
 * @brief Deleting a node in the linked list
 * array
 * @date 2025-03-04
 *
 * @copyright Copyright (c) 2025
'''
import sys

def Node(data,next):
    return [data, next]

def insertNodeInList(head, data, position):
    temp1 = Node(data,None)
    if (1 == position):
        temp1[1] = head
        head = temp1
    else:
        temp2 = head
        for i in range(0,position-2):
            temp2 = temp2[1]
        temp1[1] = temp2[1]
        temp2[1] = temp1
    return head

def deleteNodeInList(head, position):
    temp1 = head
    if (1 == position):
        head = temp1[1]
    else:
        for i in range(0,position-2):
            temp1 = temp1[1]
        temp2 = temp1[1]
        temp1[1] = temp2[1]
    return head

def printList():
    temp = head
    print("The list is: ", end = " ")
    while (temp != None):
        print (temp[0], end = " ")
        temp = temp[1]
    print()

head = None  # Empty linked list
head = insertNodeInList(head, 9,1)
head = insertNodeInList(head, 8,2)
head = insertNodeInList(head, 7,3)
head = insertNodeInList(head, 6,4)
head = insertNodeInList(head, 5,5)
printList()
position = int(input("Enter a position to be deleted: "))
head = deleteNodeInList(head,position)
printList()